### DISCUSSION TITLE
A mechanism to filter Rule application

### DISCUSSION THREAD

**Discussion by peterfox** [rectorphp/rector#8430] [Ideas]
Something I think would be useful is to have the ability to apply rules but in a more limited scope.

For instance. Maybe I have a rule to make classes Final, but I don't want to apply it to all classes but classes extending a certain other class. But I would like to enforce that all classes that extend another particular class are made Final. I could enforce this by PHPStan or another tool but ultimately a more useful situation would be to have Rector do the work automatically. At current, I would have to do either:

1. Create a rule specifically for that kind of class with the changes I want to apply.
2. Create a rule for making Classes Final and then give it a configuration of what ones can.

I'm thinking in the config we could have:

```php
$config->scopeRules(new OnlyExtends([\Some\Class::class]), function ($config) {
    $config->addRule(MakeClassFinal::class);
});
```

One of the things I feel is missing a bit from working on many of the Laravel rules I've created is configuration can be difficult as it's hard to predict new and interesting ways Rules could be used. It works for Upgrading projects but less so for regular integration.

**Comment by TomasVotruba** [comment]
Hi, that sounds interested and I'd use it myself. E.g. to finalize all controllers or add return type declaratoins there.

The config registration might not work, as rule is a service registered in the whole project.
Instead it would require some filtering mechanism on a `Scope $scope` + `Class_` node most likely, but could be doable.

Could you come up with a POC PR?

**Reply by peterfox** [reply]
I will try. Not sure how I'd start on this one.
